tests/test_bootstrap.py

```diff
@@ -0,0 +1,551 @@
+
+        .. note:: Here we need to reset `self.ctx.bootstrap.distribution`
+                  because, otherwise we will be unable to successfully pass the
+                  test :meth:`TestBootstrapBasic.test_attributes`
+        """
+        if self.ctx.bootstrap:
+            self.ctx.bootstrap.distribution = None
+        self.ctx.bootstrap = None
+
+
+class TestBootstrapBasic(BaseClassSetupBootstrap, unittest.TestCase):
+    """
+    An inherited class of `BaseClassSetupBootstrap` and `unittest.TestCase`
+    which will be used to perform tests for the methods/attributes shared
+    between all bootstraps which inherits from class
+    :class:`~pythonforandroid.bootstrap.Bootstrap`.
+    """
+
+    def test_properties(self):
+        """A test which will initialize a bootstrap and will check if the
+        values are the expected.
+        """
+        bs = Bootstrap().get_bootstrap("sdl2", self.ctx)
+        self.assertEqual(bs.name, "sdl2")
+        self.assertEqual(bs.jni_dir, "sdl2/jni")
+        self.assertEqual(bs.get_build_dir_name(), "sdl2-python3")
+
+        # test dist_dir error
+        with self.assertRaises(SystemExit) as e:
+            getattr(bs, "dist_dir")
+        self.assertEqual(e.exception.args[0], 1)
+
+        # test dist_dir success
+        self.setUp_distribution_with_bootstrap(bs)
+        self.assertTrue(bs.dist_dir.endswith("dists/test_prj"))
+
+    def test_build_dist_dirs(self):
+        """A test which will initialize a bootstrap and will check if the
+        directories we set has the values that we expect. Her we test methods:
+
+            - :meth:`~pythonforandroid.bootstrap.Bootstrap.get_build_dir`
+            - :meth:`~pythonforandroid.bootstrap.Bootstrap.get_dist_dir`
+            - :meth:`~pythonforandroid.bootstrap.Bootstrap.get_common_dir`
+        """
+        bs = Bootstrap().get_bootstrap("sdl2", self.ctx)
+        self.setUp_distribution_with_bootstrap(bs)
+
+        self.assertTrue(
+            bs.get_build_dir().endswith("build/bootstrap_builds/sdl2-python3")
+        )
+        self.assertTrue(bs.get_dist_dir("test_prj").endswith("dists/test_prj"))
+        self.assertTrue(
+            bs.get_common_dir().endswith("pythonforandroid/bootstraps/common")
+        )
+
+    def test_list_bootstraps(self):
+        """A test which will initialize a bootstrap and will check if the
+        method :meth:`~pythonforandroid.bootstrap.Bootstrap.list_bootstraps`
+        returns the expected values, which should be: `empty", `service_only`,
+        `webview` and `sdl2`
+        """
+        # prepare bootstrap and distribution
+        bs = Bootstrap().get_bootstrap("sdl2", self.ctx)
+        self.setUp_distribution_with_bootstrap(bs)
+
+        expected_bootstraps = {"empty", "service_only", "webview", "sdl2"}
+        set_of_bootstraps = set(Bootstrap().list_bootstraps())
+        self.assertEqual(
+            expected_bootstraps, expected_bootstraps & set_of_bootstraps
+        )
```

Comment:
Why are we intersecting expected_bootstraps & set_of_bootstraps?
Is it because it's OK to have `set_of_bootstraps = expected_bootstraps + {'whatever_other_bootstrap'}` while not OK to have less than `expected_bootstraps` in `set_of_bootstraps`? But then the the assertEqual in the next line defeats that purpose no?

Reply:
Yea, the main idea is that, in case that we add a new bootstrap and we don't update this test, the alarm should be triggered. Notice that the first test will only contemplate a bootstrap removal or rename, but not an addition...so the second test fills this gap.